### PR TITLE
add run_command

### PR DESCRIPTION
**Pull request by dbasunag**

### CODE REVIEW

**Comment by redhat-qe-bot1** [comment]
Report bugs in [Issues](https://github.com/myakove/github-webhook-server/issues)

The following are automatically added:
 * Add reviewers from OWNER file (in the root of the repository) under reviewers section.
 * Set PR size label.
 * New issue is created for the PR. (Closed when PR is merged/closed)
 * Run [pre-commit](https://pre-commit.ci/) if `.pre-commit-config.yaml` exists in the repo.

Available user actions:
 * To mark PR as WIP comment `/wip` to the PR, To remove it from the PR comment `/wip cancel` to the PR.
 * To block merging of PR comment `/hold`, To un-block merging of PR comment `/hold cancel`.
 * To mark PR as verified comment `/verified` to the PR, to un-verify comment `/verified cancel` to the PR.
        verified label removed on each new commit push.
 * To cherry pick a merged PR comment `/cherry-pick <target branch to cherry-pick to>` in the PR.
    * Multiple target branches can be cherry-picked, separated by spaces. (`/cherry-pick branch1 branch2`)
    * Cherry-pick will be started when PR is merged
 * To build and push container image command `/build-and-push-container` in the PR (tag will be the PR number).
 * To add a label by comment use `/<label name>`, to remove, use `/<label name> cancel`
 * To assign reviewers based on OWNERS file use `/assign-reviewers`
 * To check if PR can be merged use `/check-can-merge`

<details>
<summary>Supported /retest check runs</summary>

 * `/retest tox`: Retest tox
 * `/retest python-module-install`: Retest python-module-install

</details>

<details>
<summary>Supported labels</summary>

 * hold
 * verified
 * wip
 * lgtm

</details>

**Comment by redhat-qe-bot2** [comment]
Label verify is not a predefined one, will not be added / removed.
Available labels:

 * hold
 * verified
 * wip
 * lgtm

File: pyhelper_utils/utils.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import subprocess
```
Comment (myakove) [inline pyhelper_utils/utils.py:1]: This all repo is one big utils, we need to find another name for the file (not utils.py).
commands, executes etc...

**Review by myakove — CHANGES_REQUESTED** [review]

File: pyhelper_utils/utils.py (hunk 1 above)
Comment (myakove) [reply pyhelper_utils/utils.py:1]: `shell` is a good name for this file.

File: pyhelper_utils/utils.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+from simple_logger.logger import get_logger
+
+
+LOGGER = get_logger(name=__name__)
+
+TIMEOUT_30MIN = 30 * 60
+
+
+def run_command(
```
Comment (myakove) [inline pyhelper_utils/utils.py:11]: please add typing.

File: pyhelper_utils/utils.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        timeout (int, optional): Command wait timeout
+        capture_output (bool, default False): Capture command output
+        check (bool, default True):  If check is True and the exit code was non-zero, it raises a
+            CalledProcessError
+        hide_log_command (bool, default False): If hide_log_command is True and check will be set to False,
+            CalledProcessError will not get raise and command will not be printed.
+
+    Returns:
+        tuple: True, out if command succeeded, False, err otherwise.
+    """
```
Comment (myakove) [inline pyhelper_utils/utils.py:37]: Add raises to docstring, when check is `True`, `run` will raise `CalledProcessError` on error.

File: pyhelper_utils/utils.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    command_for_log = ["Hide", "By", "User"] if hide_log_command else command
+
+    LOGGER.info(f"Running {' '.join(command_for_log)} command")
+
+    sub_process = subprocess.run(
+        command,
+        capture_output=capture_output,
+        check=check if not hide_log_command else False,
```
Comment (myakove) [inline pyhelper_utils/utils.py:45]: We need to add comment that explain this logic. @rnetser

**Review by myakove — CHANGES_REQUESTED** [review]

File: tests/test_utils.py [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+@patch("pyhelper_utils.utils.subprocess.run")
+def test_run_command_error(mock_run):
+    mock_out = MagicMock()
+    mock_out.configure_mock(**{"stdout": "", "stderr": FAILURE_MESSAGE, "returncode": 0})
+
+    mock_run.return_value = mock_out
+    rc, out, error = run_command(command=shlex.split("true"), capture_output=False, check=False, shell=True)
+    assert not rc, ERROR_MESSAGE.format(expected="0", actual="1")
+    assert FAILURE_MESSAGE in error, ERROR_MESSAGE.format(expected=FAILURE_MESSAGE, actual="error")
+    assert not out, ERROR_MESSAGE.format(expected="", actual=out)
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_utils.py:39]: It is the condition when you have rc = 0 but have stderr. You may or may not have stdout (in my case I mocked, so I know there should be no stdout. (code path line 62 of utils.py)

File: tests/test_utils.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    with pytest.raises(CalledProcessError):
+        run_command(command=shlex.split("false"), check=True, verify_stderr=False)
+
+
+@patch("pyhelper_utils.utils.subprocess.run")
+def test_run_command_error(mock_run):
+    mock_out = MagicMock()
+    mock_out.configure_mock(**{"stdout": "", "stderr": FAILURE_MESSAGE, "returncode": 0})
+
+    mock_run.return_value = mock_out
```
Comment (rnetser) [inline tests/test_utils.py:35]:
```suggestion
def test_run_command_error(mocker):
    mocker.patch("pyhelper_utils.utils.subprocess.run", return_value=....)
```

**Review by rnetser — CHANGES_REQUESTED** [review]

File: tests/test_utils.py (hunk 6 above)
Comment (dbasunag) [reply tests/test_utils.py:35]: This requires getting extra package pytest-mock, while unittest does not need any new installations. I feel without much differences between the two, we can go with unittest and MagicMock() being straightforward and powerful, would be very helpful with run_ssh_command related tests

File: pyhelper_utils/utils.py (hunk 1 above)
Comment (dbasunag) [reply pyhelper_utils/utils.py]: done.

File: tests/test_utils.py [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    rc, out, error = run_command(command=shlex.split("true"), capture_output=False, check=False, shell=True)
+    assert not rc, ERROR_MESSAGE.format(expected="0", actual="1")
```
[the remark this answers is not in the document]
Comment (dbasunag) [reply tests/test_utils.py]: done

File: pyhelper_utils/utils.py (hunk 2 above)
Comment (dbasunag) [reply pyhelper_utils/utils.py]: done.

File: pyhelper_utils/utils.py (hunk 3 above)
Comment (dbasunag) [reply pyhelper_utils/utils.py]: Done, but this part is already mentioned in the doc string, under `check`

File: pyhelper_utils/utils.py (hunk 4 above)
Comment (dbasunag) [reply pyhelper_utils/utils.py]: Not done, as I need more information on this one. I did not find adequate information from the PR that added it.

File: pyhelper_utils/shell.py [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+        check (bool, default True):  If check is True and the exit code was non-zero, it raises a
+            CalledProcessError
+        hide_log_command (bool, default False): If hide_log_command is True and check will be set to False,
+            CalledProcessError will not get raise and command will not be printed.
+
+    Returns:
+        tuple: True, out if command succeeded, False, err otherwise.
+
+    Raises:
+        CalledProcessError when check is True and command execution fails
```
Comment (myakove) [inline pyhelper_utils/shell.py:38]:
```suggestion
        CalledProcessError: when check is True and command execution fails
```

File: tox.ini [hunk 9]
```diff
@@ -10,4 +10,4 @@ deps =
     poetry
 commands =
     poetry install
-    poetry run pytest tests
+    poetry run pytest tests -svv
```
Comment (myakove) [inline tox.ini:13]: please don't `-svv`

**Review by myakove — CHANGES_REQUESTED** [review]

File: tox.ini (hunk 9 above)
Comment (dbasunag) [reply tox.ini]: Done.

File: tests/test_utils.py (hunk 6 above)
Comment (myakove) [reply tests/test_utils.py]: pytest mock using unittest mock, so same power for both.
I think we need to stay with pytest since this is our test infra.

**Review by myakove — DISMISSED** [review]

**Review by rnetser — APPROVED** [review]

**Review by rnetser — APPROVED** [review]

**Review by myakove — APPROVED** [review]

**Pull request merged**